File: 백준/Silver/13565. 침투/침투.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

N, M = map(int, input().split())
arr = []
for _ in range(N):
    arr.append(list(map(int, input().strip())))
answer = "NO"

def dfs(start, end):
    global answer
    stack = [(start, end)]

    while stack:
        x, y = stack.pop()
        if x == N-1:
            answer = "YES"
            return True
        if visited[x][y] == 0 and arr[x][y] == 0:
            visited[x][y] = 1

            for n, m in [[0, 1], [0, -1], [1, 0], [-1, 0]]:
                nx = n + x
                my = m + y
                if 0 <= nx < N and 0 <= my < M:
                    if arr[nx][my] == 0:
                        stack.append((nx, my))

visited = [[0 for _ in range(M)] for _ in range(N)]
for x in range(M):
    if dfs(0, x):
        break
print(answer)

File: 백준/Silver/13565. 침투/침투.js
let fs = require('fs')
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const [N, M] = input[0].split(" ").map(x => parseInt(x))
let arr = []
for (let i = 1; i < input.length; i++) {
    arr.push([...input[i].trim()].map(x => parseInt(x)))
}
visited = Array.from({length : N}, () => Array(M).fill(0))

function dfs(n, m) {
    let stack = [[n, m]]

    while (stack.length) {

        const p = stack.pop()
        let x = p[0]
        let y = p[1]
        if (x === N-1) {
            return true
        }

        if (arr[x][y] === 0 && visited[x][y] === 0) {
            visited[x][y] = 1
            const dir = [[0, 1], [0, -1], [1, 0], [-1, 0]]
            for (d of dir) {
                xd = x + d[0]
                yd = y + d[1]
                
                if (xd >= 0 && xd < N && yd >= 0 && yd < M) {
                    if (visited[xd][yd] == 0 && arr[xd][yd] == 0) {
                        stack.push([xd, yd])
                    }
                }
            }
        }
    }
}

answer = "NO"
for(let i = 0; i < M; i++) {
    if (dfs(0, i)) {
        answer = "YES"
        break
    }
}
console.log(answer)
